### PR TITLE
Adversarial sample

### PR DESCRIPTION
**Pull request by gx-wind**
We come from Baidu security lab. We want to contribute the advbox tools which can generate adversarial sample to fool the neural networks like tensorflow's cleverhans library.

### CODE REVIEW

File: adversarial/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Advbox
```
Comment (reyoung) [inline adversarial/README.md]: `/adversarial/`  may not be a good path. However, we can just merge this PR, and change it later.

File: adversarial/advbox/attacks/base.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+"""
+The base model of the model.
+"""
+from abc import ABCMeta
+#from advbox.base import Model
```
Comment (reyoung) [inline adversarial/advbox/attacks/base.py]: Either uncomment or remove this line.

File: adversarial/advbox/attacks/base.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
+"""
+The base model of the model.
+"""
+from abc import ABCMeta
```
Comment (reyoung) [inline adversarial/advbox/attacks/base.py]:
```python
from abc import ABCMeta, abstractmethod
```

File: adversarial/README.md (hunk 1 above)
Comment (gx-wind) [reply adversarial/README.md]: OK. You can specify an appropriate directory. Thank you.

File: adversarial/advbox/attacks/base.py (hunk 2 above)
Comment (gx-wind) [reply adversarial/advbox/attacks/base.py]: This line has been deleted.

File: adversarial/advbox/models/paddle.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+class PaddleModel(Model):
+    """
+    Create a PaddleModel instance.
+    When you need to generate a adversarial sample, you should construct an instance of PaddleModel.
+
+    Args:
+        program(paddle.v2.fluid.framework.Program): The program of the model which generate the adversarial sample.
+        input_name(string): The name of the input.
+        logits_name(string): The name of the logits.
```
[the remark this answers is not in the document]
Comment (gx-wind) [reply adversarial/advbox/models/paddle.py]: adversarial attack has two types, untargeted and targeted.
* The goal of the non-targeted attack is to slightly modify source image in a way that image will be classified incorrectly
* The goal of the targeted attack is to slightly modify source image in a way that image will be classified as specified target class

In untargeted attack which we has implemented, the fake label logits make no effect.
In targeted attack, the logits_name means the fake label logits to cheat NN.We will implement the targeted attack later.

File: adversarial/advbox/models/paddle.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        self._input_name = input_name
+        self._logits_name = logits_name
+        self._predict_name = predict_name
+        self._cost_name = cost_name
+
+        # gradient
+        loss = self._program.block(0).var(self._cost_name)
+        param_grads = fluid.backward.append_backward(
+            loss, parameter_list=[self._input_name])
+        self._gradient = param_grads[0][1]
```
[the remark this answers is not in the document]
Comment (gx-wind) [reply adversarial/advbox/models/paddle.py]: has fixed.

File: adversarial/fluid_mnist.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    """
+    Mnist cnn model
+
+    Args:
+        img(Varaible): the input image to be recognized
+
+    Returns:
+        Variable: the label prediction
+    """
+    #conv1 = fluid.nets.conv2d()
```
[the remark this answers is not in the document]
Comment (gx-wind) [reply adversarial/fluid_mnist.py]: has removed

File: adversarial/fluid_mnist.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                                fetch_list=[avg_cost] + accuracy.metrics)
+            pass_acc = accuracy.eval(exe)
+            print("pass_id=" + str(pass_id) + " acc=" + str(acc) + " pass_acc="
+                  + str(pass_acc))
+            # print loss, acc
+            if loss < LOSS_THRESHOLD and pass_acc > ACC_THRESHOLD:
+                # if avg cost less than 10.0 and accuracy is larger than 0.9, we think our code is good.
+                break
+
+#                exit(0)
```
[the remark this answers is not in the document]
Comment (gx-wind) [reply adversarial/fluid_mnist.py]: has removed.

File: adversarial/fluid_mnist.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+        pass_acc = accuracy.eval(exe)
+        print("pass_id=" + str(pass_id) + " pass_acc=" + str(pass_acc))
+    fluid.io.save_params(
+        exe, dirname='./mnist', main_program=fluid.default_main_program())
+    print('train mnist done')
+    exit(1)
```
[the remark this answers is not in the document]
Comment (gx-wind) [reply adversarial/fluid_mnist.py]: has removed.

File: adversarial/advbox/attacks/gradientsign.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+        min_, max_ = self.model.bounds()
+        gradient = self.model.gradient(image_batch)
+        gradient_sign = np.sign(gradient) * (max_ - min_)
+
+        if not isinstance(epsilons, Iterable):
+            epsilons = np.linspace(0, 1, num=epsilons + 1)
+
+        for epsilon in epsilons:
+            adv_img = image_batch[0][0].reshape(
```
[the remark this answers is not in the document]
Comment (gx-wind) [reply adversarial/advbox/attacks/gradientsign.py]: Yes. I have changed the name and modify the comment to avoid ambiguity.

File: adversarial/advbox/models/paddle.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (JiayiFeng) [reply adversarial/advbox/models/paddle.py]: In my understanding, even in the non-targeted attack, the fake label logits are required to generate the loss. Otherwise, we can neither get the backward gradient nor modify input images further. However, what the fake label logits are is not important. They can be all zeros, random vectors or even different each batch.

Do I understand it correctly?

**Comment by wangkuiyi** [comment]
I noticed that this PR merges code into the root directory. Let's create an examples directory, and move the code there. I created an issue to remark this https://github.com/PaddlePaddle/Paddle/issues/7542

File: adversarial/advbox/models/paddle.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gx-wind) [reply adversarial/advbox/models/paddle.py]: yes
